File: macOS/Sources/CodexAccountBar/AuthTokenRefreshService.swift
import Foundation

struct AuthTokenRefreshService: Sendable {
    private let tokenURL = URL(string: "https://auth.openai.com/oauth/token")!
    private let clientID = "app_EMoamEEZ73f0CkXaXp7hrann"

    func refresh(authData: Data) async throws -> Data {
        guard let root = try JSONSerialization.jsonObject(with: authData) as? [String: Any],
              let tokens = root["tokens"] as? [String: Any],
              let refreshToken = tokens["refresh_token"] as? String,
              !refreshToken.isEmpty
        else {
            throw AuthTokenRefreshError.missingRefreshToken
        }

        var components = URLComponents()
        components.queryItems = [
            URLQueryItem(name: "grant_type", value: "refresh_token"),
            URLQueryItem(name: "client_id", value: clientID),
            URLQueryItem(name: "refresh_token", value: refreshToken)
        ]
        var request = URLRequest(url: tokenURL)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.httpBody = components.percentEncodedQuery?.data(using: .utf8)
        request.timeoutInterval = 20

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
            throw AuthTokenRefreshError.requestFailed((response as? HTTPURLResponse)?.statusCode ?? 0)
        }
        let refreshed = try JSONDecoder().decode(OAuthRefreshResponse.self, from: data)
        return try AuthDocumentUpdater.applyingRefresh(
            to: authData,
            accessToken: refreshed.accessToken,
            idToken: refreshed.idToken,
            refreshToken: refreshed.refreshToken,
            refreshedAt: Date()
        )
    }
}

enum AuthDocumentUpdater {
    static func applyingRefresh(
        to authData: Data,
        accessToken: String,
        idToken: String?,
        refreshToken: String?,
        refreshedAt: Date
    ) throws -> Data {
        guard var root = try JSONSerialization.jsonObject(with: authData) as? [String: Any],
              var tokens = root["tokens"] as? [String: Any]
        else {
            throw AuthTokenRefreshError.invalidAuth
        }
        tokens["access_token"] = accessToken
        if let idToken, !idToken.isEmpty { tokens["id_token"] = idToken }
        if let refreshToken, !refreshToken.isEmpty { tokens["refresh_token"] = refreshToken }
        root["tokens"] = tokens
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        root["last_refresh"] = formatter.string(from: refreshedAt)
        return try JSONSerialization.data(withJSONObject: root, options: [.sortedKeys])
    }
}

private struct OAuthRefreshResponse: Decodable {
    let accessToken: String
    let idToken: String?
    let refreshToken: String?

    enum CodingKeys: String, CodingKey {
        case accessToken = "access_token"
        case idToken = "id_token"
        case refreshToken = "refresh_token"
    }
}

enum AuthTokenRefreshError: LocalizedError {
    case missingRefreshToken
    case invalidAuth
    case requestFailed(Int)

    var errorDescription: String? {
        switch self {
        case .missingRefreshToken: "This account must be switched once to renew its sign-in."
        case .invalidAuth: "The saved Codex auth document is invalid."
        case .requestFailed(let status): "Codex sign-in refresh failed (HTTP \(status))."
        }
    }
}

File: Services/AuthTokenRefreshService.cs
using System.Net.Http;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace CodexAccountBar.Services;

public sealed class AuthTokenRefreshService
{
    private const string ClientId = "app_EMoamEEZ73f0CkXaXp7hrann";
    private readonly HttpClient _http = new() { Timeout = TimeSpan.FromSeconds(20) };

    public async Task<string> RefreshAsync(string authJson)
    {
        var root = JsonNode.Parse(authJson)?.AsObject()
            ?? throw new InvalidDataException("The saved Codex auth document is invalid.");
        var tokens = root["tokens"]?.AsObject()
            ?? throw new InvalidDataException("The saved Codex auth document is invalid.");
        var refreshToken = tokens["refresh_token"]?.GetValue<string>();
        if (string.IsNullOrWhiteSpace(refreshToken))
        {
            throw new InvalidDataException("This account must be switched once to renew its sign-in.");
        }

        using var form = new FormUrlEncodedContent(new Dictionary<string, string>
        {
            ["grant_type"] = "refresh_token",
            ["client_id"] = ClientId,
            ["refresh_token"] = refreshToken
        });
        using var response = await _http.PostAsync("https://auth.openai.com/oauth/token", form);
        var responseJson = await response.Content.ReadAsStringAsync();
        if (!response.IsSuccessStatusCode)
        {
            throw new HttpRequestException(
                $"Codex sign-in refresh failed (HTTP {(int)response.StatusCode}).",
                null,
                response.StatusCode
            );
        }

        using var document = JsonDocument.Parse(responseJson);
        var payload = document.RootElement;
        tokens["access_token"] = RequiredString(payload, "access_token");
        if (OptionalString(payload, "id_token") is { Length: > 0 } idToken)
        {
            tokens["id_token"] = idToken;
        }
        if (OptionalString(payload, "refresh_token") is { Length: > 0 } rotatedRefreshToken)
        {
            tokens["refresh_token"] = rotatedRefreshToken;
        }
        root["last_refresh"] = DateTimeOffset.UtcNow.ToString("O");
        return root.ToJsonString();
    }

    private static string RequiredString(JsonElement root, string name) =>
        OptionalString(root, name)
        ?? throw new InvalidDataException($"Codex sign-in refresh response is missing {name}.");

    private static string? OptionalString(JsonElement root, string name) =>
        root.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.String
            ? value.GetString()
            : null;
}
